### PR TITLE
allow * to be used in registry path for remove-item

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
Code is creating a DirectoryInfo instance to determine if the path is a directory or file, but it fails because `*` is not allowed, however, `*` is a valid character in a registry path.  Fix is to catch this exception and ignore it.  This means that using `-literalpath` with an asterisk and the filesystem provider will now return an error rather than returning quietly.

Fix https://github.com/PowerShell/PowerShell/issues/4704
Fix https://github.com/PowerShell/PowerShell/issues/4330

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Navigation.cs [hunk 1]
```diff
@@ -3266,6 +3266,10 @@ protected override void ProcessRecord()
                     {
                         // not a directory
                     }
+                    catch (System.ArgumentException)
```
Comment (anmenaga) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Navigation.cs]: Maybe it is better to skip that <DirectoryInfo code section> altogether if provider is not FileSystem provider?
e.g.: resolvedPath.Provider.Name.Equals(FileSystemProvider.ProviderName, StringComparison.OrdinalIgnoreCase)

Comment (anmenaga) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Navigation.cs]: ... that <DirectoryInfo code section> section altogether ...

Comment (anmenaga) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Navigation.cs]: ... that "DirectoryInfo code section" section altogether ...

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Navigation.cs [hunk 2]
```diff
@@ -3266,6 +3266,10 @@ protected override void ProcessRecord()
...
+                    {
```
Comment (anmenaga) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Navigation.cs]: Need to make sure that FileSystem paths continue to work as they were before this new catch handler; 2 tests:
1) remove-item with -Path and filesystem path that contains *
2) remove-item with -LiteralPath and filesystem path that contains *

**Review by anmenaga — CHANGES_REQUESTED** [review]

File: test/powershell/Modules/Microsoft.PowerShell.Management/Registry.Tests.ps1 [hunk 3]
```diff
@@ -469,6 +469,22 @@ Describe "Extended Registry Provider Tests" -Tags @("Feature", "RequireAdminOnWi
...
+            try {
+                $tempPath = "HKCU:\_tmp"
+                $testPath = "$tempPath\*\sub"
+                New-Item -Force $testPath
+                $testPath | Should Exist
+                Remove-Item -LiteralPath $testPath
+                $testPath | Should Not Exist
+            }
+            finally {
+                Remove-Item -Recurse $tempPath
```
Comment (adityapatwardhan) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Registry.Tests.ps1]: Please add `-ErrorAction SilentlyContinue -Force`

File: test/powershell/Modules/Microsoft.PowerShell.Management/Registry.Tests.ps1 [hunk 4]
```diff
@@ -469,6 +469,22 @@ Describe "Extended Registry Provider Tests" -Tags @("Feature", "RequireAdminOnWi
             $result."$testPropertyName" | Should Be $testPropertyValue
         }
     }
+
+    Context "Validate -LiteralPath" {
+        It "Verify New-Item and Remove-Item work with asterisk" {
+            try {
+                $tempPath = "HKCU:\_tmp"
+                $testPath = "$tempPath\*\sub"
+                New-Item -Force $testPath
```
Comment (adityapatwardhan) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Registry.Tests.ps1]: Please assign to null.
```Powershell
$null = New-Item -Force $testPath
```

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Navigation.cs (hunk 1 above)
Comment (adityapatwardhan) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Navigation.cs]: I agree checking for the provider seems to be a better option.

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Navigation.cs (hunk 1 above)
Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Navigation.cs]: will make the change

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Navigation.cs (hunk 2 above)
Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Navigation.cs]: Looks like there's tests for -Path, I'll add a test for -LiteralPath with *

File: test/powershell/Modules/Microsoft.PowerShell.Management/Registry.Tests.ps1 (hunk 4 above)
Comment (SteveL-MSFT) [reply test/powershell/Modules/Microsoft.PowerShell.Management/Registry.Tests.ps1]: will fix

File: test/powershell/Modules/Microsoft.PowerShell.Management/Registry.Tests.ps1 (hunk 3 above)
Comment (SteveL-MSFT) [reply test/powershell/Modules/Microsoft.PowerShell.Management/Registry.Tests.ps1]: will add

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee reviewed this and is ok with the minor `breaking change`

**Review by adityapatwardhan — APPROVED** [review]
LGTM

**Comment by adityapatwardhan** [comment]
@anmenaga Can you have a look again?

**Review by anmenaga — APPROVED** [review]

**Comment by adityapatwardhan** [comment]
@SteveL-MSFT Can you update the PR description.

**Pull request merged**